x/staking/handler.go

```diff
@@ -229,7 +229,17 @@ func handleMsgDelegate(ctx sdk.Context, msg types.MsgDelegate, k keeper.Keeper)
 }
 
 func handleMsgUndelegate(ctx sdk.Context, msg types.MsgUndelegate, k keeper.Keeper) sdk.Result {
-	completionTime, err := k.Undelegate(ctx, msg.DelegatorAddress, msg.ValidatorAddress, msg.SharesAmount)
+	validator, found := k.GetValidator(ctx, msg.ValidatorAddress)
+	if !found {
+		return ErrNoValidatorFound(k.Codespace()).Result()
+	}
+
+	if validator.Tokens.IsZero() {
+		return ErrInsufficientShares(k.Codespace()).Result()
+	}
+
+	sharesAmt := validator.DelegatorShares.MulInt(msg.Amount.Amount).QuoInt(validator.Tokens).TruncateDec()
```

Comment:
Three things, using `TruncateDec()` here is a big nono - that will chop all the decimals points off making it impossible for the full withdrawal of shares from a validator (the decimal shares will be left in the account).

Reply:
ACKed.